GeoJSON support + GL maps

ajithranka opened the pull request:
Changes:

- Refactored to ES2015, React v15 and Redux
- Replaces Leaflet with Mapbox GL JS
- Supports the new backend api

cc: @aaronlidman @Rub21 @batpad

batpad commented on src/components/shared/Settings.js:
```diff
@@ -38,21 +30,22 @@ module.exports = React.createClass({
     return (
-      /* jshint ignore:start */
-      <div id='modal' className='animate modal modal-content active' onClick={this.onCancel}>
+      <div id='modal' className='animate modal modal-content active' onClick={closeSettings}>
```
@ajithranka is it a good idea to use an id like `modal` here if it is only for this specific modal? Seems like we could run into conflicting ids on the page?

ajithranka replied:
Fixed in [8cd1792](https://github.com/osmlab/to-fix/pull/160/commits/8cd1792727d20269801c7950b6821ec03c98f8cb).

ajithranka commented on package.json:
```diff
@@ -2,34 +2,33 @@
+    "redux": "^3.6.0",
+    "redux-logger": "^2.7.0",
+    "redux-thunk": "^2.1.0",
+    "turf": "^3.0.14",
```
Fixed in [7c81712](https://github.com/osmlab/to-fix/pull/160/commits/7c81712285d40c23398eb0fe0d0eed2224868e08).

Replacing turf and lodash with specific submodules decreased the build size by 100KB. 🚀

tmcw commented on src/components/stats/StatsGraph.js:
```diff
@@ -0,0 +1,83 @@
+    if (statsByDate && statsByDate.length) {
+      const data = this.props.statsByDate.map(function(d) {
+        d.value = 0;
+        if (d.edit && (filter === 'all' || filter === 'edited')) d.value += d.edit;
```
Given this small distinction between filter values and data values, would it clean up this code significantly to rename the filters to be exactly the same as the data values, so that you could reference directly?

Additionally, if the default value is `0`, avoiding its addition won't change the output of the method.

For instance, something like this might work (untested code):

```js
const includedFields = (filter === 'all') ?
  ['edit', 'fixed', 'skip', 'noterror'] : [filter];
d.value = includedFields.reduce((sum, field) => sum + d[field], 0);
```

tmcw commented on src/actions/index.js:
```diff
@@ -0,0 +1,134 @@
+
+// Settings
+export const toggleSidebar = () => ({
+  type: 'settings/TOGGLE_SIDEBAR'
```
action types are a good thing to store as constants, since they mean that you're relying on an exact string value at two places - action construction (here) and in the store, so it's a place where strings can quietly get refactored out of correspondence and cause the application to fail silently

tmcw commented on src/reducers/tasks.js:
```diff
@@ -0,0 +1,97 @@
+
+const isTasksAction = /^tasks/;
+
+const isRequest = /_REQUEST$/;
```
instead of using a regex and naming pattern, maybe add a property to action objects that explicitly identifies them as requests, success, or failure?

tmcw commented on src/components/admin/AddTask.js:
```diff
@@ -0,0 +1,103 @@
+  getFormData = () => {
+    const formData = new window.FormData();
+
+    formData.append('name', this.refs.name.value.trim());
```
Setting refs per-input to get values is not optimal here: we tend to use refs as a last resort when we absolutely need to interact with the DOM. Form values are best stored in state, using onChange listeners, and you have a variety of options for creating abstractions that let that be non-boilerplatey

tmcw commented on src/utils/d3Graph.js:
```diff
@@ -42,7 +40,6 @@ module.exports = {
     // TODO This is dirty
```
This is a little dirty :) and I wouldn't fault you for leaving it as-is because it gets the job done and the alternative is a bunch of work. I'm definitely in favor of [keeping DOM manipulation in React land](http://www.macwright.org/2016/10/11/d3-and-react.html) as much as possible, and in this case the big gotcha with that approach would be `d3.brush`, which gives you a lot of functionality for free. You'd have to change interactions or clone [the technique d3.brush does](https://github.com/d3/d3-brush/blob/master/src/brush.js) in a React-friendly way, which would be tricky.

tmcw commented on src/components/stats/StatsGraph.js:
```diff
@@ -0,0 +1,83 @@
+    const { filter } = this.state;
+    const { statsByDate } = this.props;
+
+    if (statsByDate && statsByDate.length) {
```
This conditional makes me a little uneasy: I can understand `statsByDate` being `[]`, but should the code also allow for its value to be not an Array? In cases like this, it's nice to have React's PropTypes to make things explicit - is something required to be an array and it can be an empty array to indicate 'no data', or is it an optional property - meaning a parent component might not even provide the prop?

ajithranka commented on src/components/stats/StatsGraph.js:
```diff
@@ -0,0 +1,83 @@
+import React, { Component } from 'react';
```
Removed the casing after [1a09522](https://github.com/osmlab/to-fix/pull/160/commits/1a095220eb4c62e3458c09a8ce0abe4b67428bef), [7af5d9c](https://github.com/osmlab/to-fix/pull/160/commits/7af5d9c3c14726bdef4e450d76128dc007364e6f) and [bf8c7da](https://github.com/osmlab/to-fix/pull/160/commits/bf8c7dafdd432e9f772e67dde677c03686cb1eac).

ajithranka commented:
Some refactors made per @tmcw's review:

- Removed casing from file names. Removed underscore prefixing in variable names.
- Split things further into separate constants, action creator, reducer, selector files. See: [constants/](https://github.com/osmlab/to-fix/tree/development/src/constants) and [stores/](https://github.com/osmlab/to-fix/tree/development/src/stores).
- Async actions are handled in one place by [async_action](https://github.com/osmlab/to-fix/blob/development/src/stores/async_action.js). Loading is indicated by a single key in the state. Any async action can [dispatch loading start/stop actions optionally](https://github.com/osmlab/to-fix/blob/development/src/stores/async_action.js#L35). A failed async action dispatches the error modal. Success modals can be shown optionally by components (see: [add_task](https://github.com/osmlab/to-fix/blob/development/src/components/admin/add_task.js#L57-L63), [edit_task](https://github.com/osmlab/to-fix/blob/development/src/components/admin/edit_task.js#L45-L50)). This removes all the regex checks from reducers.
- Removed `refs` from forms. The form state is managed by the components instead. See: [add_task](https://github.com/osmlab/to-fix/blob/development/src/components/admin/add_task.js), [edit_task](https://github.com/osmlab/to-fix/blob/development/src/components/admin/edit_task.js).
- The redux store is initialized with settings [persisted from local storage](https://github.com/osmlab/to-fix/blob/development/src/stores/store.js#L38-L46). Any changes to settings in the store is [saved back to local storage](https://github.com/osmlab/to-fix/blob/development/src/stores/store.js#L56-L67). This removes local storage dependency from the reducers.
- Added PropTypes to all components and cleaned up type checks from inside the components.

What's still broken:

- The d3 code for the brush graph. I'm not very familiar with d3 to hit this out right now :-/ Planning to take this up with @geohacker's help a little later.
- The [component](https://github.com/osmlab/to-fix/blob/development/src/components/task/index.js) that displays the GL map is quite a bit hacky.

@tmcw Thank you so much for the review. The code feels much better now.

Please let me know if you have any further suggestions.

tmcw commented:
👍 Looks great.

> The component that displays the GL map is quite a bit hacky.

A lot of that is in manipulation of the style document, which tends to be a hack since the basic API of GL JS is very stateful whereas 'the React way' emphasizes having one way to render any change. For that, you might want to generate or modify a style document directly in JavaScript, and if you need high performance, to use [the style-spec's diffing algorithm](https://github.com/mapbox/mapbox-gl-style-spec/blob/mb-pages/lib/diff.js). With those changes, we can do updates like

```js
changes.forEach(function(change) {
  map[change.command].apply(map, change.args);
});
```

But the non-optimized - just calling setStyle - method might also be fast enough, and calling setStyle always tends to make React-related code easier to manage.

ajithranka commented:
I created a base `master` branch that's even with the current `gh-pages` branch.

Going to close this PR and create a new one that merges with `master`.
